torch/_inductor/runtime/benchmarking.py

```diff
@@ -0,0 +1,944 @@
+
+        _callable()
+        torch.cuda.synchronize()
+
+        buffer = torch.empty(self.L2_cache_size // 4, dtype=torch.int, device="cuda")
+        buffer.zero_()
+
+        event_pairs = self.get_event_pairs(estimation_iters)
+        start_time_s = perf_counter()
+        for start_event, end_event in event_pairs:
+            buffer.zero_()
+            start_event.record()
+            _callable()
+            end_event.record()
+        # before we synchronize we want to measure the cpu-side cost of our buffer
+        # zero, launching the callable, and the associated event records
+        cpu_launch_overhead_ms_per_iter = (
+            (perf_counter() - start_time_s) * MILLISECONDS_PER_SECOND
+        ) / estimation_iters
+        torch.cuda.synchronize()
+        estimated_timing_ms = self.get_min_timing_ms(event_pairs)
+
+        # we don't want to benchmark for longer than max_benchmark_duration milliseconds,
+        # so we reduce benchmark_iters to fit within this constraint if necessary
+        benchmark_iters = self.reduce_benchmark_iters_to_max_benchmark_duration_ms(
+            memory_warmup_iters,
+            benchmark_iters,
+            cpu_launch_overhead_ms_per_iter,
+            estimated_timing_ms,
+            1,
+            max_benchmark_duration_ms,
+        )
+        # the self.gpu_queue_limit'th event queued to the GPU will execute synchronously
+        # on the CPU; this is bad since we try to overlap the CPU launch period with a
+        # GPU sleep to avoid the CPU launch overhead in our GPU event timings. what ends
+        # up happening in this case is that the self.gpu_queue_limit'th event will wait
+        # until the queue shrinks (i.e. after the GPU sleep ends) before queueing itself.
+        # this has two issues, first we will waste a lot of time needlesly as the CPU
+        # waits for the GPU to finish its sleep, and second we will no longer have a
+        # proper overlap between the CPU launch overhead and the GPU sleep which can
+        # result in incorrect timing values for particularly small kernels as the CPU
+        # will eventually catch up to the GPU and we will end up in lock-step execution.
+        # we can prevent this by benchmarking in blocks. we calculate the number of blocks
+        # required by assuming that each benchmark iteration spawns approximately 5 events:
+        # 1 for the buffer zeroing, 1 for the callable, 2 for the event records, and 1
+        # extra allocated for any sub-kernels the callable may launch of which the most
+        # common case is allocating an output Tensor. we then evenly split the total
+        # number of benchmark iterations across all blocks.
+        benchmark_blocks = ((benchmark_iters * 5) // (self.gpu_queue_limit - 1)) + 1
+        benchmark_iters_per_block = benchmark_iters // benchmark_blocks
+        memory_warmup_iters_per_block = memory_warmup_iters // benchmark_blocks
```

Comment:
cc @ipiszy (who added do_bench_profiling) and @int3 (who I believed looked into something simliar) for comment on this section